### PR TITLE
[Security] Fixed roles serialization on token from user object

### PR DESCRIPTION
**Pull request by eko**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14274 |
| License | MIT |
| Doc PR | - |

This PR fixes the serialization of tokens when using `Role` objects provided from the user. Indeed, there were actually a reference issue that can causes fatal errors like the following one:

```
FatalErrorException in RoleHierarchy.php line 43:
Error: Call to a member function getRole() on string
```

Here is a small code example to reproduce and its output:

``` php
$user = new Symfony\Component\Security\Core\User\User('name', 'password', [
    new Symfony\Component\Security\Core\Role\Role('name')
]);
$token = new Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken($user, 'password', 'providerKey', $user->getRoles());

$serialized = serialize($token);
$unserialized = unserialize($serialized);

var_dump($unserialized->getRoles());
```

Before:

```
array(1) { [0]=> bool(true) }
```

After:

```
array(1) { [0]=> object(Symfony\Component\Security\Core\Role\Role)#15 (1) {["role":"Symfony\Component\Security\Core\Role\Role":private]=> string(4) "name" } }
```

Thank you

### CODE REVIEW

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php [hunk 1]
```diff
@@ -46,7 +46,7 @@ public function __construct(array $roles = array())
                 throw new \InvalidArgumentException(sprintf('$roles must be an array of strings, or RoleInterface instances, but got %s.', gettype($role)));
             }
 
-            $this->roles[] = $role;
+            $this->roles[] = clone $role;
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: Why not do the cloning in `serialize()` like the user object?

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php [hunk 2]
```diff
@@ -96,6 +97,19 @@ public function testSerialize()
         $this->assertEquals($token->getAttributes(), $uToken->getAttributes());
     }
 
+    public function testSerializeWithRoleObjects()
+    {
+        $user = new User('name', 'password', array(new Role('ROLE_FOO')));
+        $token = new ConcreteToken($user, $user->getRoles());
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: Should be
```php
$roles = array(new Role('ROLE_FOO'));
$token = $this->getToken($roles);
// serialize/unserialzie
// assert against $roles
```

File: src/Symfony/Component/Security/Core/Tests/Authentication/Provider/UserAuthenticationProviderTest.php [hunk 3]
```diff
@@ -220,7 +220,7 @@ public function testAuthenticateWithPreservingRoleSwitchUserRole()
         $this->assertInstanceOf('Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken', $authToken);
         $this->assertSame($user, $authToken->getUser());
         $this->assertContains(new Role('ROLE_FOO'), $authToken->getRoles(), '', false, false);
-        $this->assertContains($switchUserRole, $authToken->getRoles());
+        $this->assertContains($switchUserRole, $authToken->getRoles(), '', false, false);
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Provider/UserAuthenticationProviderTest.php]: Is this related?

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php [hunk 4]
```diff
@@ -96,6 +97,19 @@ public function testSerialize()
         $this->assertEquals($token->getAttributes(), $uToken->getAttributes());
     }
 
+    public function testSerializeWithRoleObjects()
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: Perhaps better update `testSerialize` with a 2nd role as object.

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php (hunk 1 above)
Comment (eko) [inline src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: @ro0NL Because in the `serialize()` method, we have `$this->roles` which is an array, not an object (cannot be cloned).

File: src/Symfony/Component/Security/Core/Tests/Authentication/Provider/UserAuthenticationProviderTest.php (hunk 3 above)
Comment (eko) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Provider/UserAuthenticationProviderTest.php]: @ro0NL Yes, it is related, test failed without that change because we checked an exact reference. Check has to be changed like the line just before.

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php (hunk 4 above)
Comment (eko) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: @ro0NL I agree with you on this comment, I will remove all of these tests and just let one into `AbstractTokenTest` to avoid adding a lot of tests. Thanks

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php (hunk 1 above)
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: Understand.. i also meant: clone the array and each object in it.

Comment (eko) [inline src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: @ro0NL Alright, this is updated according to your comment.

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php (hunk 4 above)
Comment (eko) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: This is done, PR is up-to-date with comments.

Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: Why not update `testSerialize`? Ie. [L91](https://github.com/symfony/symfony/pull/19778/files#diff-c8afeb16bbcfc3913e3eaee9dfcdda3bR91) => `$token = $this->getToken(array('ROLE_FOO', new Role('ROLE_BAR'));`

Comment (eko) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: @ro0NL I've updated the `testSerialize` test with a second role that is from user entity (even if it were already working in this case).
The issue here was when the roles are the same reference as the ones in the user entity.

Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: Your change is covered by the test above now.

> The issue here was when the roles are the same reference as the ones in the user entity.

Im not sure how that relates to your change... what about adding a `$token->setUser($mockedUser)` to above test so `serialize()` is fully covered.

Comment (eko) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: @ro0NL Roles are passed in the `AbstractToken::__construct()` when coming from the user: https://github.com/eko/symfony/blob/dfa7f5020e1274870ee815bb536d37b0e61d8046/src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php#L40

**Comment by eko** [comment]
@ro0NL Thank you for taking time to review this PR, is it ok for you?

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php [hunk 5]
```diff
@@ -87,7 +88,7 @@ public function testEraseCredentials()
 
     public function testSerialize()
     {
-        $token = $this->getToken(array('ROLE_FOO'));
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: What about

```php
$user = new User('name', 'password', array(new Role('ROLE_FOO'), new Role('ROLE_BAR')));
$token = $this->getToken(array('ROLE_FOO2', new Role('ROLE_BAR2')));
$token->setUser($user);

// serialize/unserialize

$this->assertEquals($token->getUser(), $uToken->getUser());
```

Fully covers `AbstractToken` and no need for the test below.

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php [hunk 6]
```diff
@@ -96,6 +97,19 @@ public function testSerialize()
         $this->assertEquals($token->getAttributes(), $uToken->getAttributes());
     }
 
+    public function testSerializeWithRoleObjects()
+    {
+        $user = new User('name', 'password', array(new Role('ROLE_FOO'), new Role('ROLE_BAR')));
+        $token = new ConcreteToken($user, $user->getRoles());
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: Point is; passing roles here is effectively the same as passing in roles via `$this->getToken($roles)`.. like above.

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php [hunk 7]
```diff
@@ -96,6 +97,19 @@ public function testSerialize()
...
+
+        $serialized = serialize($token);
+        $unserialized = unserialize($serialized);
+
+        $roles = $unserialized->getRoles();
+
+        $this->assertEquals($roles, $user->getRoles());
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/AbstractTokenTest.php]: Same test as above, ie. `$this->assertEquals($token->getRoles(), $uToken->getRoles());`

**Comment by ro0NL** [comment]
Still think tests can be combined, but other then that this is good :+1: 

Status: Reviewed

**Comment by fabpot** [comment]
Thank you @eko.
